### ISSUE TITLE
LocalOfficeManager with OpenOffice: converter performance degradation after about 60 secs or 60 conversion processes

### ISSUE THREAD

**Issue by srssrs** [sbraconnier/jodconverter#124]
Hi,
I test JodConverter with **LocalOfficeManager** with a simple conversion of  the same **office doc** file to an **odt** file multiple times.
I have installed on my MacOS the **Apache OpenOffice 4.1.6.**

The first conversion take about 1 secs to complete the process, the following ones take about 300 msec but after about 60 conversion processes the following ones take more than 30 secs or minutes!

Following a snippet of my test
     
```
 @ParameterizedTest
	@ValueSource(strings = { SAMPLE_DOC })
	public void testConvertToODT_100Times( String fileName ) throws Exception {
		try ( JODSampleConverter gen = new JODSampleConverter() ) {
			// After about 60 conversions, each conversion process during more than 30 sec or minutes ...!!!
			for ( int i = 0; i < 100; i++ ) {
				testConverter( fileName, gen, i );
			}
		}

	}
protected void testConverter( String fileName, JODSampleConverter gen, int i ) {

		StopWatch stopWatcher = new StopWatch();
		stopWatcher.start();
		log.debug( "-----Conversion {}", i );
		long currentTimeMillis = System.currentTimeMillis();

		//
		File f = gen.convertToODT( testPath.resolve( fileName ).toFile(), testOutputPath,
				"conv_" + i + "_" + currentTimeMillis );
		assertNotNull( f );
		log.debug( "-----Conversion ({}) finished in {} millisec. ", i,
				stopWatcher.getTime( TimeUnit.MILLISECONDS ) );
	}
```

The same behavior occurs when after the first conversion you wait about 60 seconds before doing another conversion.

```
@ParameterizedTest
	@ValueSource(strings = { SAMPLE_DOC })
	public void testConvertToODT_5Times_withSleep60Sec( String fileName ) throws Exception {

		try ( JODSampleConverter gen = new JODSampleConverter() ) {
			// Test multiple time (sequence) with SLEEP after each conversion
			// After the first conversion, the conversion process during more than 30 sec or minutes ...!!!
			for ( int i = 0; i < 5; i++ ) {

				testConverter( fileName, gen, i );
				log.debug( "SLEEP 60 sec" );
				Thread.sleep( 60_000 );
			}
		}

	}

```

This is my simple class to test the conversion:
```

public class JODSampleConverter implements AutoCloseable {

	private static final Logger log = LogManager.getLogger( JODSampleConverter.class );

	private static final long TASK_QUEUE_TIMEOUT_MILLISEC = 30_000;
	private static final long TASK_TIMEOUT_MILLISEC = 120_000;
	private static final long PROCESS_TIMEOUT_MILLISEC = 30_000;

	private boolean enabled;

	private LocalOfficeManager officeManager;

	public JODSampleConverter() {

		super();
		startService();
	}

	protected void startService() {

		setEnabled( false );
		try {
			officeManager = LocalOfficeManager.builder()
					.taskQueueTimeout( TASK_QUEUE_TIMEOUT_MILLISEC )
					.taskExecutionTimeout( TASK_TIMEOUT_MILLISEC )
					.processTimeout( PROCESS_TIMEOUT_MILLISEC )
					.install()
					.build();

			log.debug( "Office manager installed" );

			officeManager.start();

			log.debug( "Office manager started" );
			setEnabled( true );

		}
		catch (Exception ex) {
			log.error( "Error during openOffice connection", ex );
		}
	}

	/**
	 * Sample convert to ODT file only FIRST PAGE
	 *
	 * @param aFile
	 * @return
	 */
	public File convertToODT( File aFile, Path outputPath, String outputFileSuffix ) {

		File outputFile = null;
		try {

			// Convert to Open Office ODT file
			outputFile = outputPath.resolve( "jodTemp" + outputFileSuffix + ".odt" ).toFile();
			log.debug( "Office manager running: {}", officeManager.isRunning() );

			// Select only the first page
			final PageSelectorFilter selectorFilter = new PageSelectorFilter( 1 );

			DocumentConverter converter = LocalConverter.builder()
					.officeManager( officeManager )
					.filterChain( selectorFilter )
					.build();

			log.debug( "Converter instatiated {}", converter );

			// Convert...
			log.debug( "Convert document {} to odt file", aFile );
			converter
					.convert( new FileInputStream( aFile ) ).as(
							DefaultDocumentFormatRegistry.getFormatByExtension(
									FilenameUtils.getExtension( aFile.getName() ) ) )

					.to( outputFile )
					.execute();

			log.debug( "Converted {}", aFile );

		}
		catch (Exception e) {
			log.error( "Error converting file " + aFile, e );

		}
		finally {

		}

		return outputFile;
	}

	public void shutdown() {

		try {
			officeManager.stop();
		}
		catch (OfficeException e) {
			log.debug( "Error during shutdown office manager ", e );
		}

	}

	public boolean isEnabled() {

		return enabled;
	}

	protected void setEnabled( boolean enabled ) {

		this.enabled = enabled;
	}

	@Override
	public void close() throws Exception {

		shutdown();

	}
```
And this is the log of the **testConvertToODT_5Times_withSleep60Sec**:

> 2019-01-03 17:56:55,312 DEBUG org.test.jodtest.JODSampleConverter.startService(JODSampleConverter.java:50) [main] Office manager installed 
> 2019-01-03 17:56:59,751 DEBUG org.test.jodtest.JODSampleConverter.startService(JODSampleConverter.java:54) [main] Office manager started 
> 2019-01-03 17:56:59,753 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:88) [main] -----Conversion 0 
> 2019-01-03 17:56:59,753 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:79) [main] Office manager running: true 
> 2019-01-03 17:56:59,804 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Converter instatiated org.jodconverter.LocalConverter@434a63ab 
> 2019-01-03 17:56:59,804 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:92) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
> 2019-01-03 17:57:00,840 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:101) [main] Converted src/test/resources/testContent/sample.doc 
> 2019-01-03 17:57:00,843 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:95) [main] -----**Conversion (0) finished in 1089 millisec.**  
> 2019-01-03 17:58:00,844 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:88) [main] -----Conversion 1 
> 2019-01-03 17:58:00,845 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:79) [main] Office manager running: true 
> 2019-01-03 17:58:00,845 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Converter instatiated org.jodconverter.LocalConverter@47d9a273 
> 2019-01-03 17:58:00,846 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:92) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
> 2019-01-03 17:59:16,889 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:101) [main] Converted src/test/resources/testContent/sample.doc 
> 2019-01-03 17:59:16,889 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:95) [main] ---**--Conversion (1) finished in 76044 millisec.**

**Comment by sbraconnier** [comment]
Could you please attach a sample document for which you experienced this issue?

I won't be able to check this out this week (I'll try but no promises). I'll do it next week for sure.

**Comment by srssrs** [comment]
Sure, I attached here the sample doc


[sample.doc.zip](https://github.com/sbraconnier/jodconverter/files/2724832/sample.doc.zip)

**Comment by srssrs** [comment]
No, unfortunately I have the same degradation also without PageSelectorFilter.

**Comment by sbraconnier** [comment]
Oh really.. hmmm. Is it possible for you to test with LibreOffice, just to see is the problem is related to OpenOffice ?

**Comment by srssrs** [comment]
Yes I test with LibreOffice and it's ok the problem is only with OpenOffice

**Comment by sbraconnier** [comment]
Are you using gradle or maven ?

**Comment by srssrs** [comment]
Maven

**Comment by sbraconnier** [comment]
Could you try to replace the LibreOffice librairies that come with JODConverter by the OpenOffice ones:

```xml
<dependencies>
    <dependency>
        <groupId>org.jodconverter</groupId>
        <artifactId>jodconverter-local</artifactId>
        <version>4.2.2</version>
        <exclusions>
            <!--  JODConverter comes with libre-office libraries -->
            <exclusion>
              <groupId>org.libreoffice</groupId>
              <artifactId>juh</artifactId>
            </exclusion>
            <exclusion>
              <groupId>org.libreoffice</groupId>
              <artifactId>jurt</artifactId>
            </exclusion>
            <exclusion>
              <groupId>org.libreoffice</groupId>
              <artifactId>ridl</artifactId>
            </exclusion>
            <exclusion>
              <groupId>org.libreoffice</groupId>
              <artifactId>unoil</artifactId>
            </exclusion>
        </exclusions>    
    </dependency>
    <dependency>
        <groupId>org.openoffice</groupId>
        <artifactId>juh</artifactId>
        <version>4.1.2</version>
    </dependency>
    <dependency>
        <groupId>org.openoffice</groupId>
        <artifactId>jurt</artifactId>
        <version>4.1.2</version>
    </dependency>
    <dependency>
        <groupId>org.openoffice</groupId>
        <artifactId>ridl</artifactId>
        <version>4.1.2</version>
    </dependency>
    <dependency>
        <groupId>org.openoffice</groupId>
        <artifactId>unoil</artifactId>
        <version>4.1.2</version>
    </dependency>
    <dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-log4j12</artifactId>
        <version>1.7.25</version>
    </dependency>
</dependencies>
```

**Comment by srssrs** [comment]
In order to test your maven configuration, I reinstalled OpenOffice 4.1.6 and delete LibreOffice Application but now when running the test case I got the following exception (without updating the pom as you suggested):
```
2019-01-22 21:27:05,651 DEBUG org.test.jodtest.JODSampleConverter.startService(JODSampleConverter.java:50) [main] Office manager installed 
2019-01-22 21:27:10,033 DEBUG org.test.jodtest.JODSampleConverter.startService(JODSampleConverter.java:54) [main] Office manager started 
2019-01-22 21:27:10,033 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 21:27:10,086 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@625732 
2019-01-22 21:27:10,086 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 21:27:10,704 ERROR org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:102) [main] Error converting file src/test/resources/testContent/sample.doc to output file target/odt/jodTempconv1548188830033.png org.jodconverter.office.OfficeException: Could not store document: jodTempconv1548188830033.png; errorCode: 2074
	at org.jodconverter.task.LocalConversionTask.storeDocument(LocalConversionTask.java:169)
	at org.jodconverter.task.LocalConversionTask.execute(LocalConversionTask.java:107)
	at org.jodconverter.office.OfficeProcessManagerPoolEntry.doExecute(OfficeProcessManagerPoolEntry.java:148)
	at org.jodconverter.office.AbstractOfficeManagerPoolEntry$1.call(AbstractOfficeManagerPoolEntry.java:77)
	at org.jodconverter.office.AbstractOfficeManagerPoolEntry$1.call(AbstractOfficeManagerPoolEntry.java:73)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.sun.star.task.ErrorCodeIOException: 
	at com.sun.star.lib.uno.environments.remote.Job.remoteUnoRequestRaisedException(Job.java:173)
	at com.sun.star.lib.uno.environments.remote.Job.execute(Job.java:139)
	at com.sun.star.lib.uno.environments.remote.JobQueue.enter(JobQueue.java:334)
	at com.sun.star.lib.uno.environments.remote.JobQueue.enter(JobQueue.java:303)
	at com.sun.star.lib.uno.environments.remote.JavaThreadPool.enter(JavaThreadPool.java:87)
	at com.sun.star.lib.uno.bridges.java_remote.java_remote_bridge.sendRequest(java_remote_bridge.java:636)
	at com.sun.star.lib.uno.bridges.java_remote.ProxyFactory$Handler.request(ProxyFactory.java:146)
	at com.sun.star.lib.uno.bridges.java_remote.ProxyFactory$Handler.invoke(ProxyFactory.java:128)
	at com.sun.proxy.$Proxy50.storeToURL(Unknown Source)
	at org.jodconverter.task.LocalConversionTask.storeDocument(LocalConversionTask.java:166)
	... 8 more
```

**Comment by sbraconnier** [comment]
Googling about this error, one reason could be because you are converting a file named "sample.doc", but its content would be something else (html, odt, etc...). Is it possible that your sample.doc file is not a Microsoft Word document?

**Comment by sbraconnier** [comment]
You message is saying that you are converting to odt, but the output filename extension is png... this is odd.

**Comment by srssrs** [comment]
I think is another bug because I'll try to convert from .doc to png (only the text log "convert to odt..." was a type error). With LibreOffice the same test run correctly. But I will investigate in the next few days and I will let you know and eventually I will open a new issue.

After modified the test as original one to convert **.doc to odt** and after modified the pom as you suggested, I have always the same degradation performance problems at the conversion number 65

```
...
2019-01-22 22:01:28,721 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:100) [main] -----Conversion 64 
2019-01-22 22:01:28,721 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 22:01:28,721 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@25ce9dc4 
2019-01-22 22:01:28,722 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 22:01:29,139 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:98) [main] Converted src/test/resources/testContent/sample.doc 
2019-01-22 22:01:29,140 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:107) [main] -----Conversion (64) finished in **418 millisec.**  
2019-01-22 22:01:29,140 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:100) [main] -----Conversion 65 
2019-01-22 22:01:29,140 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 22:01:29,141 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@74ea2410 
2019-01-22 22:01:29,141 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 22:01:29,545 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:98) [main] Converted src/test/resources/testContent/sample.doc 
2019-01-22 22:01:29,545 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:107) [main] -----Conversion (65) finished in **405 millisec.**  
2019-01-22 22:01:29,546 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:100) [main] -----Conversion 66 
2019-01-22 22:01:29,546 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 22:01:29,547 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@17f62e33 
2019-01-22 22:01:29,547 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 22:02:13,485 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:98) [main] Converted src/test/resources/testContent/sample.doc 
2019-01-22 22:02:13,486 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:107) [main] -----Conversion (66) finished in 43939 millisec. 
2019-01-22 22:02:13,486 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:100) [main] -----Conversion 67 
2019-01-22 22:02:13,486 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 22:02:13,487 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@76b1e9b8 
2019-01-22 22:02:13,487 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 22:02:15,951 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:98) [main] Converted src/test/resources/testContent/sample.doc 
2019-01-22 22:02:15,951 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:107) [main] -----Conversion (67) finished in 2465 millisec.
2019-01-22 22:02:15,951 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:100) [main] -----Conversion 68 
2019-01-22 22:02:15,952 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:76) [main] Office manager running: true 
2019-01-22 22:02:15,952 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:86) [main] Converter instatiated org.jodconverter.LocalConverter@27406a17 
2019-01-22 22:02:15,952 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:89) [main] Convert document src/test/resources/testContent/sample.doc to odt file 
2019-01-22 22:02:16,355 DEBUG org.test.jodtest.JODSampleConverter.convertToODT(JODSampleConverter.java:98) [main] Converted src/test/resources/testContent/sample.doc 
2019-01-22 22:02:16,355 DEBUG org.test.jodtest.JODSampleConverterTest.testConverter(JODSampleConverterTest.java:107) [main] -----Conversion (68) finished in 403 millisec. 

```
And after the 68th conversion again the performance go wrong.

**Comment by sbraconnier** [comment]
Wow... that is pretty strange.... is it possible for you to activate the TRACE logging level for jodconverter and put back here your entire log file, or at least until the degradation problem occurs.

**Comment by srssrs** [comment]
I attached the file log with TRACE logging for jodconverter. In this case the performance degradation happens at conversion 63. 
[jodconverter.log](https://github.com/sbraconnier/jodconverter/files/2792015/jodconverter.log)

**Comment by srssrs** [comment]
OK thank you this work-around works correctly for me!

**Comment by sbraconnier** [comment]
Glad to hear that 👍

**Comment by PerfectACE** [comment]
> OK thank you this work-around works correctly for me!

so, finally how do you deal the problem.?

**Comment by ryanrupp** [comment]
I ran into a similar problem using LibreOffice specifically on OSX. It's appears to be time based though e.g. after X seconds after startup of LibreOffice (~30 seconds maybe) something happens and the performance of LibreOffice conversions degrade. Verified it's time based e.g. my original test was just single threaded going through repeated conversions of a 2 page docx to PDF that takes 200ms until ~130-150 iterations in and then performance degrades to ~3x or worse (nothing too crazy though like 60 seconds). Then I went to 1 second sleeps between conversions and I can see this start happening on the ~30th iteration (so ~30 seconds in). Then I added 5 second sleeps in between and could see it happening on the 5th iteration (~30 seconds in). So, something appears to be happening after LibreOffice has been running for more than 30 seconds. I can see the CPU utilization goes up during that time e.g. on a 12 core machine it was consuming ~15% CPU during my tests and then once that point hit it starts using ~30%. I didn't get a chance to profile it to figure out what is going on inside of LibreOffice to make this happen.

I did however then use these same tests on Linux (Ubuntu server in a container with LibreOffice installed) and couldn't reproduce the issue although the versions were not the same, on Ubuntu it was `LibreOffice 5.2.7.2`. I haven't had a chance to circle back to narrow down if this is isolated OSX or the version of LibreOffice but given the problem description here I'm leaning towards something about OSX.
